### ISSUE TITLE
events on multiple interactive sprites or events on single stage and guess coordinates?

### ISSUE THREAD

**Issue by gg4u** [pixijs/pixi.js#4201]
I am working on a graph-layout visualisation, pixi 4.5.4.

Nodes are interactive and allow to expand the graph.

For simplicity I attached interactive sprites for each node (and each node is decorated with its own texture).

I binded pan events to the stage.

The graph looks great, but soon enough on mobile will start clobbering.

How to profile if it due to:
- the number of events (soon I will have hundreds of interactive sprites)
- the size of textures
- other things... 

Would you handle all events from the stage, as example guessing coordinates of the object closer to the touch event ?
Or is not expected to be a buttleneck having multiple interactive sprites dragging around canvas ?

**Comment by ivanpopelyshev** [comment]
Clobbering how exactly?

Size of textures is not important. Number of base textures is. Use spritesheets or atlases.

> buttleneck 

butt or neck?

Not enough info on your case :(

**Comment by gg4u** [comment]
I mean that it start to be very slow : I click on nodes, or drag the stage, and re-position of elements is slow and choppy.

Size of textures is not important : good to now.
Number of base texture is:
I guess each time I do a PIXI.Sprite.fromImage('a_different_image.png') ,  a new base texture is created.

How many could the stay on the canvas, orientatively?

I cannot use sprite sheets because images are pulled from other websites.

What is an atlas?
Could not find it in reference in documentation: https://pixijs.download/v4.5.4/docs/PIXI.GraphicsRenderer.html

**Comment by ivanpopelyshev** [comment]
Not if the string is the same. "PIXI.Sprite.fromImage('a_different_image.png') " is cached :)

Yeah, culling is a good idea. Just follow the algorithm:

1. put everything into rectangular chunks (containers).
2. set "chunk.renderable =false" if its too far from the camera.

You have to think about padding for chunks, how far away chunk has to be from camera to stop being renderable.

**Comment by gg4u** [comment]
Ok, I ll try like this.

I have different graphics objects added to stage.
That is, the graphics.parent is the stage :)

how could I get the graphics position respect to stage?

**Comment by ivanpopelyshev** [comment]
"graphics.position".

But if you have multiple things inside the graphics, then you cant get it. There's no way of knowing "center of that particular circle".

However, you dont need it! If you want interaction, I have a good code for it: https://pixijs.github.io/examples/#/projection/basic.js 

Forget about projections, you just need to figure out how my interaction works.

**Comment by gg4u** [comment]
Mmm.. it is kinda complicated here :) the positions calculated for the nodes are respect to graph object, and don't know how map them respect to "global" coordinates..
I am working on this example: https://github.com/anvaka/ngraph.pixi

**Comment by ivanpopelyshev** [comment]
Use toLocal / toGlobal of PIXI.DisplayObject . I provided you with code that makes drag&drop in local coordinates even if coordinate system is strange.

**Comment by ivanpopelyshev** [comment]
You have to study the code of 

1. Transform

https://github.com/pixijs/pixi.js/blob/dev/src/core/display/TransformStatic.js

2. toGlobal/toLocal

https://github.com/pixijs/pixi.js/blob/dev/src/core/display/DisplayObject.js#L261

There's no easy way to do it, that's one of basics of any engine or rendering lib - know how Transform works in that particular lib.

**Comment by gg4u** [comment]
ok @ivanpopelyshev , thank you for the references!

**Comment by gg4u** [comment]
Ivan, thank you so much.

So, the good part was that I was able to do it ... culling, a new thing to me :)
the the so so part is that I don't perceive any particular improvement : Doh!

I say perceive beacuse I am not particularly running any other test than counting the number of taps I do and... watch what happens!

Happy to have learnt something, but maybe the culprit in performance was not that.

However I will stay positive and think that at least there have been some improvement :)

Just to check: 
my nodes are carries a thumbnail, each thumbnail loads its own texture.

I set:
` ui.thumbnail.renderable = false;
`
for topics outside a visibile area in the canvas.

Note that nodes will be still rendered, just the texture goes off.

Another possibility:
- How to check if performance issue are due to events bound to thumbnails?
- How to check how much is due to layout computation?
- Does rendering engine can be paused once a layout computation of my object is stopped (basically, I wonder if the rendering engine continues to try to calculate positions of elements, even if their position is now constant)?

Note that with layout computation without nodes, thousands of nodes can rendered.
But with images (even if thumbnail.renderable = false) and events, it reduces to a couple of hundreds (about 40 taps per session in my design).
